pkg/process/manager.go

```diff
@@ -26,14 +26,164 @@ import (
 	"github.com/containerd/nydus-snapshotter/pkg/utils/mount"
 )
 
+type DaemonStates struct {
+	mu              sync.Mutex
+	idxBySnapshotID map[string]*daemon.Daemon // index by snapshot ID
+	idxByDaemonID   map[string]*daemon.Daemon // index by ID
+	daemons         []*daemon.Daemon          // all daemon
+}
+
+func newDaemonStates() *DaemonStates {
+	return &DaemonStates{
+		idxBySnapshotID: make(map[string]*daemon.Daemon),
+		idxByDaemonID:   make(map[string]*daemon.Daemon),
+	}
+}
+
+// Return nil if the daemon is never inserted or managed,
+// otherwise returns the previously inserted daemon pointer.
+// Allowing replace an existed daemon since some fields in Daemon can change after restarting nydusd.
+func (s *DaemonStates) Add(daemon *daemon.Daemon) *daemon.Daemon {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	old, ok := s.idxByDaemonID[daemon.ID]
+
+	// TODO: No need to retain all daemons in the slice, just use the map indexed by DaemonID
+	if ok {
+		for i, d := range s.daemons {
+			if d.ID == daemon.ID {
+				s.daemons[i] = daemon
+			}
+		}
+	} else {
+		s.daemons = append(s.daemons, daemon)
+	}
+
+	s.idxByDaemonID[daemon.ID] = daemon
+	s.idxBySnapshotID[daemon.SnapshotID] = daemon
+
+	if ok {
+		return old
```

Comment:
Why do we need to return `old`? seems it doesn't to be used.

Reply:
Yes. This is just a hint to callers. It is up to the caller if they need to check the daemon is previously inserted.
By returning  old daemon pointer, caller has a chance to tell it is overwitten.